### PR TITLE
Provide specific field to accept depfiles in target class

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

This removes the error condition where a depfile is accidentally included in only one of inputs and outputs, leading to accidental missed dependencies.

### CODE REVIEW

File: packages/flutter_tools/lib/src/build_system/source.dart [hunk 1]
```diff
@@ -198,7 +198,7 @@ abstract class Source {
   /// Before the first build, the depfile is expected to be missing. Its
   /// absence is interpreted as the build needing to run. Afterwards, both
   /// input and output file hashes are updated.
-  const factory Source.depfile(String name) = _DepfileSource;
+  ///const factory Source.depfile(String name) = _DepfileSource;
```
Comment (dramosv90) [inline packages/flutter_tools/lib/src/build_system/source.dart]:
```suggestion
  /// const factory Source.depfile(String name) = _DepfileSource;
```

File: packages/flutter_tools/test/general.shard/build_system/build_system_test.dart [hunk 2]
```diff
@@ -319,8 +319,7 @@ void main() {
       fs.file('a.txt').writeAsStringSync('a');
       called += 1;
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/build_system_test.dart]:
```suggestion
      called++;
```

File: packages/flutter_tools/test/general.shard/build_system/build_system_test.dart [hunk 3]
```diff
@@ -356,8 +355,7 @@ void main() {
       }
       called += 1;
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/build_system_test.dart]:
```suggestion
      called++;
```

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart [hunk 4]
```diff
@@ -160,57 +160,52 @@ void main() {
   }));
 
   test('can resolve a missing depfile', () => testbed.run(() {
-    const Source depfile = Source.depfile('foo.d');
+    visitor.visitDepfile('foo.d');
 
-    depfile.accept(visitor);
     expect(visitor.sources, isEmpty);
     expect(visitor.containsNewDepfile, true);
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/source_test.dart]:
```suggestion
    expect(visitor.containsNewDepfile, isTrue);
```

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart [hunk 5]
```diff
@@ -160,57 +160,52 @@ void main() {
...
 
   test('can resolve a populated depfile', () => testbed.run(() {
-    const Source depfile = Source.depfile('foo.d');
     environment.buildDir.childFile('foo.d')
       .writeAsStringSync('a.dart : c.dart');
 
-    depfile.accept(visitor);
+    visitor.visitDepfile('foo.d');
     expect(visitor.sources.single.path, 'c.dart');
     expect(visitor.containsNewDepfile, false);
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/source_test.dart]:
```suggestion
    expect(visitor.containsNewDepfile, isFalse);
```

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart [hunk 6]
```diff
@@ -160,57 +160,52 @@ void main() {
...
 
     final SourceVisitor outputVisitor = SourceVisitor(environment, false);
-    depfile.accept(outputVisitor);
+    outputVisitor.visitDepfile('foo.d');
 
     expect(outputVisitor.sources.single.path, 'a.dart');
     expect(outputVisitor.containsNewDepfile, false);
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/source_test.dart]:
```suggestion
    expect(outputVisitor.containsNewDepfile, isFalse);
```

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart [hunk 7]
```diff
@@ -160,57 +160,52 @@ void main() {
...
 
   test('does not crash on completely invalid depfile', () => testbed.run(() {
-    const Source depfile = Source.depfile('foo.d');
     environment.buildDir.childFile('foo.d')
         .writeAsStringSync('hello, world');
 
-    depfile.accept(visitor);
+    visitor.visitDepfile('foo.d');
     expect(visitor.sources, isEmpty);
     expect(visitor.containsNewDepfile, false);
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/source_test.dart]:
```suggestion
    expect(visitor.containsNewDepfile, isFalse);
```

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart [hunk 8]
```diff
@@ -160,57 +160,52 @@ void main() {
...
 
   test('can parse depfile with windows paths', () => testbed.run(() {
-    const Source depfile = Source.depfile('foo.d');
     environment.buildDir.childFile('foo.d')
         .writeAsStringSync(r'a.dart: C:\\foo\\bar.txt');
 
-    depfile.accept(visitor);
+    visitor.visitDepfile('foo.d');
     expect(visitor.sources.single.path, r'C:\foo\bar.txt');
     expect(visitor.containsNewDepfile, false);
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/source_test.dart]:
```suggestion
    expect(visitor.containsNewDepfile, isFalse);
```

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart [hunk 9]
```diff
@@ -160,57 +160,52 @@ void main() {
...
 
   test('can parse depfile with spaces in paths', () => testbed.run(() {
-    const Source depfile = Source.depfile('foo.d');
     environment.buildDir.childFile('foo.d')
         .writeAsStringSync(r'a.dart: foo\ bar.txt');
 
-    depfile.accept(visitor);
+    visitor.visitDepfile('foo.d');
     expect(visitor.sources.single.path, r'foo bar.txt');
     expect(visitor.containsNewDepfile, false);
```
Comment (dramosv90) [inline packages/flutter_tools/test/general.shard/build_system/source_test.dart]:
```suggestion
    expect(visitor.containsNewDepfile, isFalse);
```

**Review by darielopper — CHANGES_REQUESTED** [review]

File: packages/flutter_tools/test/general.shard/build_system/build_system_test.dart (hunk 2 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/build_system/build_system_test.dart]: See https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer-over

File: packages/flutter_tools/test/general.shard/build_system/source_test.dart (hunk 4 above)
Comment (jonahwilliams) [reply packages/flutter_tools/test/general.shard/build_system/source_test.dart]: I generally prefer using the literal over the matcher, though we don't have a rule on way or the  other

**Review by zanderso — APPROVED** [review]
lgtm w/ question. What happens when a depfile is declared, but the action doesn't actually write one?

**Comment by jonahwilliams** [comment]
Its treated as if it were a missing input, meaning the target will rerun each time.

**Pull request merged**